[Bug]: Realtime API crashes when client terminates connection "incorrectly"

### Your current environment

I have an Nvidia GeForce RTX 3090 and I'm running this stuff within a Docker environment. I am unable to run your environment collecting scripts because it is not designed to run in a generic Python environment that does not have vLLM installed. Perhaps a fallback that would provide what information is available would have been useful here.

### 🐛 Describe the bug

I decided I would try to give Voxtral Mini 4B 2602 a shot. Apparently, it's designed for real-time streaming, unlike what I've been using. It's also advertised to be production ready with vLLM's Realtime API.

Unfortunately, it seems that whoever wrote that did not share my view of what production-ready means. First, I have to use a `nightly` image. I fail to see how that's production-ready, but okay.

To me, production ready also means that a server should not crash if a client sends a malformed or unexpected request. I'm here to report that the vLLM server crashes completely if the client terminates the Realtime API WebSocket connection in an "unexpected" way.

In order to reproduce this bug, please follow these steps:

1. You will need to run vLLM with the `mistralai/Voxtral-Mini-4B-Realtime-2602` model. I'm running it inside of a Docker container (configuration follows), but I assume the bug is also reproducible outside of it.
2. Open a WebSocket connection to the Realtime API. E.x., ws://host/v1/realtime
3. The server will send a `session.created` event. That's fine.
4. Send an `input_audio_buffer.commit` event to the server.
5. I don't have any audio for you to send yet. Do not send `input_audio_buffer.append`. I'm sorry, but I don't have any audio for you.
6. We've decided to close the WebSocket connection. But before we do, in accordance with [the documentation](https://docs.vllm.ai/en/v0.16.0/serving/openai_compatible_server/#realtime-api), please send `input_audio_buffer.commit` with `final=True` to signal we have no further audio to send. I know that we never send any audio in the first place, but it can't hurt to inform the server, right?
7. What do you mean 'the server crashed'? What do you mean with provide a `multimodal_embedding` at every step? 

Here is my docker-compose.yml:

```yaml
services:
  voxtral-realtime:
    build: .
    #image: vllm/vllm-openai:nightly
    container_name: voxtral-realtime
    runtime: nvidia
    environment:
      - NVIDIA_VISIBLE_DEVICES=all
      - HF_HOME=/root/.cache/huggingface
      - VLLM_DISABLE_COMPILE_CACHE=1
    volumes:
      - /storage/models:/root/.cache/huggingface
    ports:
      - "8001:8001"
    command: >
      mistralai/Voxtral-Mini-4B-Realtime-2602 --host 0.0.0.0 --port 8001 --dtype bfloat16 --tokenizer-mode mistral --config-format mistral --load-format mistral --trust-remote-code --compilation-config '{"cudagraph_mode":"PIECEWISE"}' --tensor-parallel-size 1 --max-model-len 16384 --max-num-seqs 1 --max-num-batched-tokens 256 --gpu-memory-utilization 0.90
```

Here is my `Dockerfile`:

```
FROM vllm/vllm-openai:nightly

RUN apt-get update && apt-get install -y --no-install-recommends libsndfile1 \
 && rm -rf /var/lib/apt/lists/*

RUN pip install --no-cache-dir "mistral-common[soundfile]"
```

Here is the full stack trace:

```
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [dump_input.py:72] Dumping input data for V1 LLM engine (v0.16.0rc2.dev165+gdcf6ee859) with config: model='mistralai/Voxtral-Mini-4B-Realtime-2602', speculative_config=None, tokenizer='mistralai/Voxtral-Mini-4B-Realtime-2602', skip_tokenizer_init=False, tokenizer_mode=mistral, revision=None, tokenizer_revision=None, trust_remote_code=True, dtype=torch.bfloat16, max_seq_len=16384, download_dir=None, load_format=mistral, tensor_parallel_size=1, pipeline_parallel_size=1, data_parallel_size=1, disable_custom_all_reduce=False, quantization=None, enforce_eager=False, enable_return_routed_experts=False, kv_cache_dtype=auto, device_config=cuda, structured_outputs_config=StructuredOutputsConfig(backend='auto', disable_fallback=False, disable_any_whitespace=False, disable_additional_properties=False, reasoning_parser='', reasoning_parser_plugin='', enable_in_reasoning=False), observability_config=ObservabilityConfig(show_hidden_metrics_for_version=None, otlp_traces_endpoint=None, collect_detailed_traces=None, kv_cache_metrics=False, kv_cache_metrics_sample=0.01, cudagraph_metrics=False, enable_layerwise_nvtx_tracing=False, enable_mfu_metrics=False, enable_mm_processor_stats=False, enable_logging_iteration_details=False), seed=0, served_model_name=mistralai/Voxtral-Mini-4B-Realtime-2602, enable_prefix_caching=True, enable_chunked_prefill=True, pooler_config=None, compilation_config={'level': None, 'mode': <CompilationMode.VLLM_COMPILE: 3>, 'debug_dump_path': None, 'cache_dir': '/root/.cache/vllm/torch_compile_cache/e2b17d18c5', 'compile_cache_save_format': 'binary', 'backend': 'inductor', 'custom_ops': ['none'], 'splitting_ops': ['vllm::unified_attention', 'vllm::unified_attention_with_output', 'vllm::unified_mla_attention', 'vllm::unified_mla_attention_with_output', 'vllm::mamba_mixer2', 'vllm::mamba_mixer', 'vllm::short_conv', 'vllm::linear_attention', 'vllm::plamo2_mamba_mixer', 'vllm::gdn_attention_core', 'vllm::kda_attention', 'vllm::sparse_attn_indexer', 'vllm::rocm_aiter_sparse_attn_indexer', 'vllm::unified_kv_cache_update'], 'compile_mm_encoder': False, 'compile_sizes': [], 'compile_ranges_split_points': [256], 'inductor_compile_config': {'enable_auto_functionalized_v2': False, 'combo_kernels': True, 'benchmark_combo_kernel': True}, 'inductor_passes': {}, 'cudagraph_mode': <CUDAGraphMode.PIECEWISE: 1>, 'cudagraph_num_of_warmups': 1, 'cudagraph_capture_sizes': [1, 2, 4, 8, 16, 24, 32], 'cudagraph_copy_inputs': False, 'cudagraph_specialize_lora': True, 'use_inductor_graph_partition': False, 'pass_config': {'fuse_norm_quant': False, 'fuse_act_quant': False, 'fuse_attn_quant': False, 'enable_sp': False, 'fuse_gemm_comms': False, 'fuse_allreduce_rms': False, 'fuse_act_padding': False}, 'max_cudagraph_capture_size': 32, 'dynamic_shapes_config': {'type': <DynamicShapesType.BACKED: 'backed'>, 'evaluate_guards': False, 'assume_32_bit_indexing': False}, 'local_cache_dir': '/root/.cache/vllm/torch_compile_cache/e2b17d18c5/rank_0_0/backbone', 'fast_moe_cold_start': True, 'static_all_moe_layers': []}, 
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [dump_input.py:79] Dumping scheduler output for model execution: SchedulerOutput(scheduled_new_reqs=[NewRequestData(req_id=rt-ws-5e748a7f-b6dd-4b92-91f9-3fb607d1a86c-87fa1a34-dc5c-4c0e-b6b2-160f061e14fa-a9eb6c75,prompt_token_ids_len=1,prefill_token_ids_len=None,mm_features=[],sampling_params=SamplingParams(n=1, presence_penalty=0.0, frequency_penalty=0.0, repetition_penalty=1.0, temperature=0.0, top_p=1.0, top_k=0, min_p=0.0, seed=None, stop=[], stop_token_ids=[], bad_words=[], include_stop_str_in_output=False, ignore_eos=False, max_tokens=1, min_tokens=0, logprobs=None, prompt_logprobs=None, skip_special_tokens=True, spaces_between_special_tokens=True, truncate_prompt_tokens=None, structured_outputs=None, extra_args=None),block_ids=([9], [10]),num_computed_tokens=0,lora_request=None,prompt_embeds_shape=None)], scheduled_cached_reqs=CachedRequestData(req_ids=[],resumed_req_ids=set(),new_token_ids_lens=[],all_token_ids_lens={},new_block_ids=[],num_computed_tokens=[],num_output_tokens=[]), num_scheduled_tokens={rt-ws-5e748a7f-b6dd-4b92-91f9-3fb607d1a86c-87fa1a34-dc5c-4c0e-b6b2-160f061e14fa-a9eb6c75: 1}, total_num_scheduled_tokens=1, scheduled_spec_decode_tokens={}, scheduled_encoder_inputs={}, num_common_prefix_blocks=[0, 0], finished_req_ids=[], free_encoder_mm_hashes=[], preempted_req_ids=[], has_structured_output_requests=false, pending_structured_output_tokens=false, num_invalid_spec_tokens=null, kv_connector_metadata=null, ec_connector_metadata=null)
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [dump_input.py:81] Dumping scheduler stats: SchedulerStats(num_running_reqs=1, num_waiting_reqs=0, step_counter=0, current_wave=0, kv_cache_usage=0.002739726027397249, encoder_cache_usage=0.0098876953125, prefix_cache_stats=PrefixCacheStats(reset=False, requests=1, queries=1, hits=0, preempted_requests=0, preempted_queries=0, preempted_hits=0), connector_prefix_cache_stats=None, kv_cache_eviction_events=[], spec_decoding_stats=None, kv_connector_stats=None, waiting_lora_adapters={}, running_lora_adapters={}, cudagraph_stats=None, perf_stats=None)
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038] EngineCore encountered a fatal error.
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038] Traceback (most recent call last):
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/engine/core.py", line 1029, in run_engine_core
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]     engine_core.run_busy_loop()
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/engine/core.py", line 1058, in run_busy_loop
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]     self._process_engine_step()
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/engine/core.py", line 1096, in _process_engine_step
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]     outputs, model_executed = self.step_fn()
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]                               ^^^^^^^^^^^^^^
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/engine/core.py", line 518, in step_with_batch_queue
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]     exec_model_fut.result()
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]   File "/usr/lib/python3.12/concurrent/futures/_base.py", line 449, in result
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]     return self.__get_result()
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]            ^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]   File "/usr/lib/python3.12/concurrent/futures/_base.py", line 401, in __get_result
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]     raise self._exception
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/executor/uniproc_executor.py", line 79, in collective_rpc
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]     result = run_method(self.driver_worker, method, args, kwargs)
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/serial_utils.py", line 459, in run_method
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]     return func(*args, **kwargs)
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]            ^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/worker/worker_base.py", line 361, in execute_model
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]     return self.worker.execute_model(scheduler_output)
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]   File "/usr/local/lib/python3.12/dist-packages/torch/utils/_contextlib.py", line 124, in decorate_context
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]     return func(*args, **kwargs)
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]            ^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/worker/gpu_worker.py", line 648, in execute_model
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]     output = self.model_runner.execute_model(
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]   File "/usr/local/lib/python3.12/dist-packages/torch/utils/_contextlib.py", line 124, in decorate_context
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]     return func(*args, **kwargs)
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]            ^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/worker/gpu_model_runner.py", line 3502, in execute_model
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]     ) = self._preprocess(
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]         ^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/worker/gpu_model_runner.py", line 2768, in _preprocess
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]     inputs_embeds_scheduled = self.model.embed_input_ids(
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]   File "/usr/local/lib/python3.12/dist-packages/vllm/model_executor/models/voxtral_realtime.py", line 305, in embed_input_ids
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]     assert len(multimodal_embeddings) > 0, (
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=84) ERROR 02-13 17:56:25 [core.py:1038] AssertionError: For realtime you must provide a multimodal_embedding at every step.
(EngineCore_DP0 pid=84) Process EngineCore_DP0:
(APIServer pid=1) ERROR 02-13 17:56:25 [async_llm.py:715] AsyncLLM output_handler failed.
(APIServer pid=1) ERROR 02-13 17:56:25 [async_llm.py:715] Traceback (most recent call last):
(APIServer pid=1) ERROR 02-13 17:56:25 [async_llm.py:715]   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/engine/async_llm.py", line 671, in output_handler
(APIServer pid=1) ERROR 02-13 17:56:25 [async_llm.py:715]     outputs = await engine_core.get_output_async()
(APIServer pid=1) ERROR 02-13 17:56:25 [async_llm.py:715]               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(APIServer pid=1) ERROR 02-13 17:56:25 [async_llm.py:715]   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/engine/core_client.py", line 913, in get_output_async
(APIServer pid=1) ERROR 02-13 17:56:25 [async_llm.py:715]     raise self._format_exception(outputs) from None
(APIServer pid=1) ERROR 02-13 17:56:25 [async_llm.py:715] vllm.v1.engine.exceptions.EngineDeadError: EngineCore encountered an issue. See stack trace (above) for the root cause.
(EngineCore_DP0 pid=84) Traceback (most recent call last):
(EngineCore_DP0 pid=84)   File "/usr/lib/python3.12/multiprocessing/process.py", line 314, in _bootstrap
(EngineCore_DP0 pid=84)     self.run()
(EngineCore_DP0 pid=84)   File "/usr/lib/python3.12/multiprocessing/process.py", line 108, in run
(EngineCore_DP0 pid=84)     self._target(*self._args, **self._kwargs)
(EngineCore_DP0 pid=84)   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/engine/core.py", line 1040, in run_engine_core
(EngineCore_DP0 pid=84)     raise e
(EngineCore_DP0 pid=84)   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/engine/core.py", line 1029, in run_engine_core
(EngineCore_DP0 pid=84)     engine_core.run_busy_loop()
(EngineCore_DP0 pid=84)   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/engine/core.py", line 1058, in run_busy_loop
(EngineCore_DP0 pid=84)     self._process_engine_step()
(EngineCore_DP0 pid=84)   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/engine/core.py", line 1096, in _process_engine_step
(EngineCore_DP0 pid=84)     outputs, model_executed = self.step_fn()
(EngineCore_DP0 pid=84)                               ^^^^^^^^^^^^^^
(EngineCore_DP0 pid=84)   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/engine/core.py", line 518, in step_with_batch_queue
(EngineCore_DP0 pid=84)     exec_model_fut.result()
(EngineCore_DP0 pid=84)   File "/usr/lib/python3.12/concurrent/futures/_base.py", line 449, in result
(EngineCore_DP0 pid=84)     return self.__get_result()
(EngineCore_DP0 pid=84)            ^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=84)   File "/usr/lib/python3.12/concurrent/futures/_base.py", line 401, in __get_result
(EngineCore_DP0 pid=84)     raise self._exception
(EngineCore_DP0 pid=84)   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/executor/uniproc_executor.py", line 79, in collective_rpc
(EngineCore_DP0 pid=84)     result = run_method(self.driver_worker, method, args, kwargs)
(EngineCore_DP0 pid=84)              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=84)   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/serial_utils.py", line 459, in run_method
(EngineCore_DP0 pid=84)     return func(*args, **kwargs)
(EngineCore_DP0 pid=84)            ^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=84)   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/worker/worker_base.py", line 361, in execute_model
(EngineCore_DP0 pid=84)     return self.worker.execute_model(scheduler_output)
(EngineCore_DP0 pid=84)            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=84)   File "/usr/local/lib/python3.12/dist-packages/torch/utils/_contextlib.py", line 124, in decorate_context
(EngineCore_DP0 pid=84)     return func(*args, **kwargs)
(EngineCore_DP0 pid=84)            ^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=84)   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/worker/gpu_worker.py", line 648, in execute_model
(EngineCore_DP0 pid=84)     output = self.model_runner.execute_model(
(EngineCore_DP0 pid=84)              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=84)   File "/usr/local/lib/python3.12/dist-packages/torch/utils/_contextlib.py", line 124, in decorate_context
(EngineCore_DP0 pid=84)     return func(*args, **kwargs)
(EngineCore_DP0 pid=84)            ^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=84)   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/worker/gpu_model_runner.py", line 3502, in execute_model
(EngineCore_DP0 pid=84)     ) = self._preprocess(
(EngineCore_DP0 pid=84)         ^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=84)   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/worker/gpu_model_runner.py", line 2768, in _preprocess
(EngineCore_DP0 pid=84)     inputs_embeds_scheduled = self.model.embed_input_ids(
(EngineCore_DP0 pid=84)                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=84)   File "/usr/local/lib/python3.12/dist-packages/vllm/model_executor/models/voxtral_realtime.py", line 305, in embed_input_ids
(EngineCore_DP0 pid=84)     assert len(multimodal_embeddings) > 0, (
(EngineCore_DP0 pid=84)            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=84) AssertionError: For realtime you must provide a multimodal_embedding at every step.
[rank0]:[W213 17:56:25.137438066 ProcessGroupNCCL.cpp:1553] Warning: WARNING: destroy_process_group() was not called before program exit, which can leak resources. For more info, please see https://pytorch.org/docs/stable/distributed.html#shutdown (function operator())
(APIServer pid=1) INFO:     Shutting down
(APIServer pid=1) INFO:     Waiting for application shutdown.
(APIServer pid=1) INFO:     Application shutdown complete.
(APIServer pid=1) INFO:     Finished server process [1]
```

The root cause is most likely in the following files:

File "/usr/local/lib/python3.12/dist-packages/vllm/model_executor/models/voxtral_realtime.py", line 305, in embed_input_ids
File "/usr/local/lib/python3.12/dist-packages/vllm/v1/worker/gpu_model_runner.py", line 2768, in _preprocess


The code for the Voxtral Realtime model appears to be throwing an assertion error which is not being caught in the GPU model runner. The most likely fix for this problem would be to catch the assertion error and terminate the WebSocket connection with an error. Alternatively, if that's not feasible, the next likely fix seems to be to ignore the commit in the event that no audio was ever appended in that session.

Obviously I have fixed this on my client side, but given that this is something that completely crashes the server, I would think that it needs to be fixed before you can really even begin to claim that Realtime API is production-ready.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

--- a/tests/entrypoints/openai/test_realtime_validation.py
+++ b/tests/entrypoints/openai/test_realtime_validation.py
@@ -121,3 +121,75 @@ async def test_multi_chunk_streaming(
                 " it sleeps with quite a flow, and everywhere that Mary went,"
                 " the lamb was sure to go."
             )
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize("model_name", [MODEL_NAME])
+async def test_empty_commit_does_not_crash_engine(
+    model_name, mary_had_lamb_audio_chunks, rocm_aiter_fa_attention
+):
+    """Test that committing without audio does not crash the engine.
+
+    Regression test for https://github.com/vllm-project/vllm/issues/34532.
+    An empty commit (no prior input_audio_buffer.append) used to trigger
+    ``AssertionError: For realtime you must provide a multimodal_embedding
+    at every step`` which killed the entire engine process, disconnecting
+    every connected client.
+    """
+    server_args = ["--enforce-eager", "--max-model-len", "2048"]
+
+    if model_name.startswith("mistralai"):
+        server_args += MISTRAL_FORMAT_ARGS
+
+    add_attention_backend(server_args, rocm_aiter_fa_attention)
+
+    with RemoteOpenAIServer(model_name, server_args) as remote_server:
+        ws_url = _get_websocket_url(remote_server)
+
+        # --- First connection: empty commit (no audio appended) ----------
+        async with websockets.connect(ws_url) as ws:
+            event = await receive_event(ws, timeout=30.0)
+            assert event["type"] == "session.created"
+
+            await send_event(ws, {"type": "session.update", "model": model_name})
+
+            # Start generation without sending any audio
+            await send_event(ws, {"type": "input_audio_buffer.commit"})
+
+            # Immediately signal end-of-audio
+            await send_event(ws, {"type": "input_audio_buffer.commit", "final": True})
+
+            # We should get *some* response (error or empty transcription),
+            # but the engine must NOT crash.
+            event = await receive_event(ws, timeout=30.0)
+            assert event["type"] in (
+                "error",
+                "transcription.done",
+                "transcription.delta",
+            )
+
+        # --- Second connection: normal transcription ---------------------
+        # Verifies the engine is still alive after the empty commit above.
+        async with websockets.connect(ws_url) as ws:
+            event = await receive_event(ws, timeout=30.0)
+            assert event["type"] == "session.created"
+
+            await send_event(ws, {"type": "session.update", "model": model_name})
+
+            await send_event(ws, {"type": "input_audio_buffer.commit"})
+
+            for chunk in mary_had_lamb_audio_chunks:
+                await send_event(
+                    ws, {"type": "input_audio_buffer.append", "audio": chunk}
+                )
+
+            await send_event(ws, {"type": "input_audio_buffer.commit", "final": True})
+
+            done_received = False
+            while not done_received:
+                event = await receive_event(ws, timeout=60.0)
+                if event["type"] == "transcription.done":
+                    done_received = True
+                elif event["type"] == "error":
+                    pytest.fail(f"Engine error after empty commit: {event}")
+            assert done_received

--- a/vllm/model_executor/models/voxtral_realtime.py
+++ b/vllm/model_executor/models/voxtral_realtime.py
@@ -299,13 +299,29 @@ class VoxtralRealtimeGeneration(VoxtralForConditionalGeneration, SupportsRealtim
         # Multi-modal token ID may exceed vocab size
         handle_oov_mm_token: bool = True,
     ) -> torch.Tensor:
-        """Pass post-conv embeddings directly as input"""
-        # for realtime we simply flatten the multimodal embeddings
-        # to be in tensor format, we treat the input ids later
-        assert multimodal_embeddings is not None
-        assert len(multimodal_embeddings) > 0, (
-            "For realtime you must provide a multimodal_embedding at every step."
-        )
+        """Pass post-conv embeddings directly as input.
+
+        For realtime models, multimodal embeddings are required at every
+        decode step.  If they are missing (e.g. due to an empty audio
+        commit, encoder-cache eviction under GPU memory pressure, or a
+        client disconnect), return zero embeddings instead of crashing
+        the engine so that all other in-flight requests stay alive.
+        """
+        if multimodal_embeddings is None or len(multimodal_embeddings) == 0:
+            logger.warning(
+                "Realtime model received empty multimodal embeddings "
+                "for %d input tokens. Returning zero embeddings to "
+                "avoid engine crash.",
+                input_ids.shape[0],
+            )
+            pool_size = self.config.audio_config.block_pool_size
+            embed_dim = self.config.audio_config.d_model * pool_size
+            return torch.zeros(
+                input_ids.shape[0],
+                embed_dim,
+                dtype=self.whisper_encoder.dtype,
+                device=input_ids.device,
+            )
         mm_embeds_flat = _flatten_embeddings(multimodal_embeddings)
         return mm_embeds_flat
 
@@ -367,9 +383,12 @@ class VoxtralRealtimeGeneration(VoxtralForConditionalGeneration, SupportsRealtim
         """Transform audio waveforms -> initial whisper post-conv embeddings"""
         audio_inputs = self._parse_and_validate_audio_arrays(**kwargs)
 
-        assert audio_inputs is not None, (
-            "For realtime you must provide an audio input at every step."
-        )
+        if audio_inputs is None:
+            logger.warning(
+                "Realtime model received no audio inputs in "
+                "embed_multimodal. Returning empty embeddings."
+            )
+            return []
 
         def _truncate_left(
             sample: torch.Tensor, mult_of: int, pos: int

